{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test file Index 1\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Good quality score\n",
    "CCTTCGAC\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Low quality score\n",
    "CCTTCGAC\n",
    "+\n",
    "########\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with no Ns\n",
    "TTTTTTTT\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Ns\n",
    "NTTTTTTT\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key in dictionary \n",
    "AGAGTCCA\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key in dictionary \n",
    "AGAGTCCA\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key not in dictionary \n",
    "GGGGGGGG\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key not in dictionary \n",
    "CCCCCCCC\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key in dictionary and match \n",
    "AGGATAGC\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key in dictionary and match \n",
    "AGGATAGC\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key in dictionary, match, bad QS \n",
    "AGGATAGC\n",
    "+\n",
    "########\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key in dictionary and match, diff barcode\n",
    "TACCGGAT\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key in dictionary and dont match \n",
    "TACCGGAT\n",
    "+\n",
    "JJJJJJJJ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test file Index 1\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read with low quality score\n",
    "CCTTCGAC\n",
    "+\n",
    "########\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read with good quality score\n",
    "CCTTCGAC\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read with N\n",
    "NTTTTTTT\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read with no N\n",
    "TTTTTTTT\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read not in dictionary\n",
    "GGGGGGGG\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read not in dictionary\n",
    "TTTTTTTT\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read in dictionary\n",
    "TACGCTAC\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read  in dictionary\n",
    "TACGCTAC\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read  in dictionary and match\n",
    "GCTATCCT\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read  in dictionary and match\n",
    "GCTATCCT\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-read  in dictionary, match, bad QS\n",
    "GCTATCCT\n",
    "+\n",
    "########\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key not in dictionary and match, diff barcode \n",
    "ATCCGGTA\n",
    "+\n",
    "JJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1-reads with Key in dictionary and dont match \n",
    "CGCATGAT\n",
    "+\n",
    "JJJJJJJJ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test file R1\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "1)Good quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "2)Bad quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "3)No Ns\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "4)Has Ns\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "5)Key in dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "6)Key in dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "7)Key not in dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "8)Key not in dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "9)Key in dictionary and matches \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "10)Key in dictionary and matches\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "11)Key in dictionary and matches, bad quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "12)Key in dictionary and matches, Good quality score, and match\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "13)Key in dictionary, Good quality score, and dont match\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test file R2\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "1)Bad quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "2)Good quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "3)Has Ns\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "4)No Ns\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "5)Key not in Dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "6)Key not in Dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "7)Key in Dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "8)Key in Dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "9)Key in dictionary and matches \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "10)Key in dictionary and matches\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "11)Key in dictionary and matches, bad quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "12)Key in dictionary and matches, Good quality score, and match\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1\n",
    "13)Key in dictionary, Good quality score, and dont match\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Barcode file, tab sperated\n",
    "B1\tGTAGCGTA\n",
    "A5\tCGATCGAT\n",
    "C1\tGATCAAGG\n",
    "B9\tAACAGCGA\n",
    "C9\tTAGCCATG\n",
    "C3\tCGGTAATC\n",
    "B3\tCTCTGGAT\n",
    "C4\tTACCGGAT\n",
    "A11\tCTAGCTCA\n",
    "C7\tCACTTCAC\n",
    "B2\tGCTACTCT\n",
    "A1\tACGATCAG\n",
    "B7\tTATGGCAC\n",
    "A3\tTGTTCCGT\n",
    "B4\tGTCCTAAG\n",
    "A12\tTCGACAAG\n",
    "C10\tTCTTCGAC\n",
    "A2\tATCATGCG\n",
    "C2\tATCGTGGT\n",
    "A10\tTCGAGAGT\n",
    "B8\tTCGGATTC\n",
    "A7\tGATCTTGC\n",
    "B10\tAGAGTCCA\n",
    "A8\tAGGATAGC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Demultiplexing Python Assignment - developing code\n",
    "##################################################################################################################\n",
    "##Making Dictionary for Barcodes and output files\n",
    "\n",
    "#dictionary for R1 barcodes for making output file\n",
    "file_dict_r1={}\n",
    "#dictionary for R2 barcodes for making output files\n",
    "file_dict_r2={}\n",
    "#file that contains the barcodes and samples \n",
    "barcodes_file=\"/Users/mandiedriskill/Bi624_Assignments/demultiplexing/Barcodes.txt\"\n",
    "#dictionary for checking if index matches\n",
    "barcode_dict = {}\n",
    "#Reading each line in the barcode file\n",
    "for line in open(barcodes_file).readlines():\n",
    "    #striping the lines and spliting in the barcode file\n",
    "    sample, barcode = line.strip().split()\n",
    "    #setting the barcodes keys and the samples as values \n",
    "    barcode_dict[barcode] = sample\n",
    "    #creating R1 files that are named with all the barcodes in the dictionary\n",
    "    file_dict_r1[barcode]=open(barcode + \"_R1.fq\", \"w\")\n",
    "    #creating R2 files that are named with all the barcodes in the dictionary\n",
    "    file_dict_r2[barcode]=open(barcode + \"_R2.fq\", \"w\")\n",
    "\n",
    "\n",
    "##################################################################################################################\n",
    "##Creating Functions\n",
    "\n",
    "#making a function called convert_phred \n",
    "def convert_phred(letter):\n",
    "    #creating a variable that can be called later that converts the letter to a phred score\n",
    "    phred_score= ord(letter)-33\n",
    "    #returning the phred_score\n",
    "    return(phred_score)\n",
    "\n",
    "#making a function called reverse_complement\n",
    "def reverse_complement(bases):\n",
    "    #a dictionary that has the compliment of each base as values\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N':'N'}\n",
    "    #returning the compliment base and making the reverse.\n",
    "    return ''.join([complement[base] for base in bases[::-1]])\n",
    "\n",
    "##################################################################################################################\n",
    "#Creating Counters \n",
    "\n",
    "line_ctr=0\n",
    "\n",
    "cutoff=30\n",
    "\n",
    "ctr_low_qs=0\n",
    "ctr_good_qs=0\n",
    "\n",
    "ctr_yes_N=0 \n",
    "ctr_No_N=0\n",
    "\n",
    "barcodes_notin_dict=0\n",
    "barcodes_in_dict=0\n",
    "\n",
    "\n",
    "reads_No_match=0\n",
    "reads_Do_match=0\n",
    "\n",
    "##################################################################################################################\n",
    "##Creating Variables for Files and Reading in Files\n",
    "\n",
    "#setting variables for each file\n",
    "r1=\"/Users/mandiedriskill/Bi624_Assignments/demultiplexing/test_r1.txt\"\n",
    "r2=\"/Users/mandiedriskill/Bi624_Assignments/demultiplexing/test_r2.txt\"\n",
    "i1=\"/Users/mandiedriskill/Bi624_Assignments/demultiplexing/test_i1.txt\"\n",
    "i2=\"/Users/mandiedriskill/Bi624_Assignments/demultiplexing/test_i2.txt\"\n",
    "#opening R1,R2,I1,I2 files and setting as variables\n",
    "#creating R1 and R2 undetermined files and setting as variables\n",
    "#creating a count data file and setting as variables\n",
    "with \\\n",
    "open(r1, \"r\") as r_1, open(r2, \"r\") as r_2, \\\n",
    "open(i1, \"r\") as i_1, open(i2, \"r\") as i_2, \\\n",
    "open('R1_undetermine.fq', 'w') as undetermined_r1, \\\n",
    "open('R2_undetermine.fq', 'w') as undetermined_r2, \\\n",
    "open('count_data', 'w') as count_data:\n",
    "    \n",
    "##################################################################################################################\n",
    "##Creating Variables for Individual Lines\n",
    "    \n",
    "    while r_1 and r_2 and i_1 and i_2:\n",
    "        #setting each line as a variable and stripping the new ling at the end\n",
    "        line1_i1 = i_1.readline().strip('\\n')\n",
    "        if not line1_i1:\n",
    "            break\n",
    "        line2_i1 = i_1.readline().strip('\\n')\n",
    "        line3_i1 = i_1.readline().strip('\\n')\n",
    "        line4_i1 = i_1.readline().strip('\\n')      \n",
    "        \n",
    "        #setting each line as a variable and stripping the new ling at the end\n",
    "        line1_i2 = i_2.readline().strip('\\n')\n",
    "        if not line1_i2:\n",
    "            break\n",
    "        line2_i2 = i_2.readline().strip('\\n')\n",
    "        line3_i2 = i_2.readline().strip('\\n')\n",
    "        line4_i2 = i_2.readline().strip('\\n')\n",
    "        \n",
    "        #setting each line as a variable and stripping the new ling at the end\n",
    "        line1_r1 = r_1.readline().strip('\\n')\n",
    "        if not line1_r1:\n",
    "            break\n",
    "        line2_r1 = r_1.readline().strip('\\n')\n",
    "        line3_r1 = r_1.readline().strip('\\n')\n",
    "        line4_r1 = r_1.readline().strip('\\n')\n",
    "        #creating a variable that prints all 4 lines in fastq format with the index in the header\n",
    "        lines_r1 = line1_r1+\":\"+line2_i1+'\\n'+line2_r1+'\\n'+line3_r1+'\\n'+line4_r1+'\\n'\n",
    "        #print(lines_r1)\n",
    "        \n",
    "        #setting each line as a variable and stripping the new ling at the end\n",
    "        line1_r2 = r_2.readline().strip('\\n')\n",
    "        if not line1_r2:\n",
    "            break\n",
    "        line2_r2 = r_2.readline().strip('\\n')\n",
    "        line3_r2 = r_2.readline().strip('\\n')\n",
    "        line4_r2 = r_2.readline().strip('\\n')\n",
    "        \n",
    "        rev_i2 = reverse_complement(line2_i2)\n",
    "        #creating a variable that prints all 4 lines in fastq format with the index in the header\n",
    "        lines_r2 = line1_r2+\":\"+rev_i2+'\\n'+line2_r2+'\\n'+line3_r2+'\\n'+line4_r2+'\\n'\n",
    "        #print(line2_i2)\n",
    "        #print(lines_r2)\n",
    "\n",
    "##################################################################################################################        \n",
    "##Low Average Quality Score Removal \n",
    "\n",
    "#removing average read quality scores that are below a cutoff of 30. I choose 30 because all the quality \n",
    "##score were above 30 in the demultiplex index graphs, which any score above 30 is 99.9% accurate. In the \n",
    "##demultiplex index graphs, any cutoff above 30 will remove a good portion of good reads and we don't want\n",
    "##to do that. Low read quality scores are most likely due to Ns, which will be removed, so what will be left \n",
    "##is reads with average quality scores above 30 (99.9% accurate).  \n",
    "\n",
    "    \n",
    "        #creating a list to store characters in for index1   \n",
    "        phred_list_i1=[]\n",
    "        #looking at each character in the quality score line for index1\n",
    "        for char in line4_i1:\n",
    "            #converting each character to a phred score\n",
    "            phred_score_i1=(convert_phred(char))\n",
    "            #print(phred_score_i1)#checking right characters are printed\n",
    "            \n",
    "        #creating a list to store characters in for index2    \n",
    "        phred_list_i2=[]\n",
    "        #looking at each character in the quality score line for index2\n",
    "        for char in line4_i2:\n",
    "            #converting each character to a phred score\n",
    "            phred_score_i2=(convert_phred(char))\n",
    "            #print(phred_score_i2)#checking right characters are printed\n",
    "            \n",
    "            #appending characters to a list\n",
    "            phred_list_i1.append(phred_score_i1)\n",
    "            phred_list_i2.append(phred_score_i2)\n",
    "            #print(phred_list_i1)#making sure list look correctly\n",
    "            #print(phred_list_i2)#making sure list look correctly\n",
    "        \n",
    "        #creating variables that has the sum of each line\n",
    "        phred_sum_i1 = sum(phred_list_i1)\n",
    "        phred_sum_i2 = sum(phred_list_i2)\n",
    "        #print(\"1\",phred_sum_i1)#checking to make sure sums are correct\n",
    "        #print(\"2\",phred_sum_i2)#checking to make sure sums are correct\n",
    "\n",
    "        #creating variables that are the average of each index line\n",
    "        phred_ave_i1 = (phred_sum_i1/len(line4_i1))\n",
    "        phred_ave_i2 = (phred_sum_i2/len(line4_i2))\n",
    "        #print(\"1\",phred_ave_i1)#checking to make sure average are correct\n",
    "        #print(\"2\",phred_ave_i2)#checking to make sure average are correct\n",
    "\n",
    "        #writing reads to undetermined file (R1 and R2) if they are below the set cutoff level.\n",
    "        if phred_ave_i1 < cutoff or phred_ave_i2 < cutoff:\n",
    "            #counting number of reads that are below the cutoff level\n",
    "            ctr_low_qs +=1\n",
    "            undetermined_r1.write(lines_r1)\n",
    "            undetermined_r2.write(lines_r2)\n",
    "        #if reads are above the cutoff level\n",
    "        else: \n",
    "            #counting the number of reads that are above the cutoff level\n",
    "            ctr_good_qs +=1\n",
    "#################################################################################################################           \n",
    "##Removing Reads with Ns\n",
    "\n",
    "            #if index1 and index2 sequence lines have Ns. Reads are witten to undetermined files (R1 and R2)\n",
    "            if \"N\" in line2_i1 or \"N\" in line2_i2:\n",
    "                #counting number of reads that have Ns\n",
    "                ctr_yes_N +=1\n",
    "                undetermined_r1.write(lines_r1)\n",
    "                undetermined_r2.write(lines_r2)\n",
    "            else:\n",
    "                #counting number of reads that don't have Ns\n",
    "                ctr_No_N +=1\n",
    "                #making the reverse compliment of index2 sequence\n",
    "                reverse_i2 = reverse_complement(line2_i2)\n",
    "                #print(\"original\", line2_i2)#viewing origional index2\n",
    "                #print(\"reverse\", reverse_i2)#making sure origional was reversed complimented\n",
    "                #finding indexes that are not in the dictionary and writing them to undertermined files (R1 and R2)\n",
    "                if line2_i1 not in barcode_dict.keys() and reverse_i2 not in barcode_dict.keys():\n",
    "                    #print(\"1\",line2_i1)#viewing index1 \n",
    "                    #print(\"2\",reverse_i2)#making sure indexes are reversed complimented\n",
    "                    #counting the indexes that are not in the dictionary\n",
    "                    barcodes_notin_dict+=1\n",
    "                    undetermined_r1.write(lines_r1)\n",
    "                    undetermined_r2.write(lines_r2)\n",
    "                \n",
    "                else:\n",
    "                    #finding the indexes that are in the dictionary\n",
    "                    if line2_i1 in barcode_dict.keys() and reverse_i2 in barcode_dict.keys():\n",
    "                        #counting the indexes that are in the dicitonary\n",
    "                        barcodes_in_dict += 1\n",
    "                        #print(\"R1\",line2_i1)#checking indexes that are found in the dictionary\n",
    "                        #print(\"R2\",reverse_i2)#checking indexes that are found in the dictionary\n",
    "                    \n",
    "                    #finind index reads that don't match and writing them to undertermined files (R1 and R2)\n",
    "                    if reverse_i2 != line2_i1:\n",
    "                        #counting index reads that don't match\n",
    "                        reads_No_match+=1\n",
    "                        #print(\"R1\",line2_i1)#checking indexes that don't match \n",
    "                        #print(\"R2\",reverse_i2)#checking indexes that don't match \n",
    "                        undetermined_r1.write(lines_r1) \n",
    "                        undetermined_r2.write(lines_r2)\n",
    "                    \n",
    "                    #finding indexes that match and if they do writing them to files that match their index sequence\n",
    "                    if reverse_i2==line2_i1:\n",
    "                        #counting reads that do match \n",
    "                        reads_Do_match+=1\n",
    "                        #print(\"R1\",line2_i1)#checking index reads that match \n",
    "                        #print(\"R2\",reverse_i2)#checking index reads that match \n",
    "                        #print(lines_r1)#checking associated reads that match \n",
    "                        #print(lines_r2)#checking associated reads that match\n",
    "                        file_dict_r1[line2_i1].write(lines_r1)\n",
    "                        file_dict_r2[reverse_i2].write(lines_r2)\n",
    "                        \n",
    "    #writing count data to a file called count_data\n",
    "    count_data.write(\"{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n\".format(\"# of Read Pairs with Low SQ R1\", ctr_low_qs, \\\n",
    "                                                               \"# of Read Pairs with Good SQ R1\", ctr_good_qs, \\\n",
    "                                                               \"# of Read Pairs with Ns\", ctr_yes_N, \\\n",
    "                                                               \"# of Read Pairs without Ns\", ctr_No_N, \\\n",
    "                                                               \"# of Read Pairs Not in Dictionary\", barcodes_notin_dict, \\\n",
    "                                                               \"# of Read Pairs in Dictionary\", barcodes_in_dict, \\\n",
    "                                                               \"# of Read Pairs that Don't Match\", reads_No_match, \\\n",
    "                                                               \"# of Read Pairs that Do Match\", reads_Do_match)) \n",
    "\n",
    "#closing files that were created from the dictionary barcodes\n",
    "for barcode in file_dict_r1:\n",
    "    file_dict_r1[barcode].close()\n",
    "    file_dict_r2[barcode].close()\n",
    "            \n",
    "\n",
    "\n",
    "         \n",
    "\n",
    "                        \n",
    "                        \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test output R1_undetermine.fq output\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:CCTTCGAC\n",
    "1)Good quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:CCTTCGAC\n",
    "2)Bad quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:TTTTTTTT\n",
    "3)No Ns\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:NTTTTTTT\n",
    "4)Has Ns\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:AGAGTCCA\n",
    "5)Key in dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:AGAGTCCA\n",
    "6)Key in dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:GGGGGGGG\n",
    "7)Key not in dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:CCCCCCCC\n",
    "8)Key not in dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:AGGATAGC\n",
    "11)Key in dictionary and matches, bad quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:TACCGGAT\n",
    "13)Key in dictionary, Good quality score, and don't match\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test output R2_undetermine.fq output\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:CCTTCGAC\n",
    "1)Bad quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:CCTTCGAC\n",
    "2)Good quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:NTTTTTTT\n",
    "3)Has Ns\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:TTTTTTTT\n",
    "4)No Ns\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:GGGGGGGG\n",
    "5)Key not in Dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:TTTTTTTT\n",
    "6)Key not in Dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:TACGCTAC\n",
    "7)Key in Dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:TACGCTAC\n",
    "8)Key in Dictionary \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:GCTATCCT\n",
    "11)Key in dictionary and matches, bad quality score\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:CGCATGAT\n",
    "13)Key in dictionary, Good quality score, and don't match\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#output files, code makes them all, but they are filled only if those reads are present.\n",
    "AACAGCGA_R1.fq\n",
    "AACAGCGA_R2.fq\n",
    "ACGATCAG_R1.fq\n",
    "ACGATCAG_R2.fq\n",
    "AGAGTCCA_R1.fq\n",
    "AGAGTCCA_R2.fq\n",
    "AGGATAGC_R1.fq\n",
    "AGGATAGC_R2.fq\n",
    "ATCATGCG_R1.fq\n",
    "ATCATGCG_R2.fq\n",
    "ATCGTGGT_R1.fq\n",
    "ATCGTGGT_R2.fq\n",
    "GTCCTAAG_R1.fq\n",
    "GTCCTAAG_R2.fq\n",
    "TACCGGAT_R1.fq\n",
    "TACCGGAT_R2.fq\n",
    "TAGCCATG_R1.fq\n",
    "TAGCCATG_R2.fq\n",
    "TATGGCAC_R1.fq\n",
    "TATGGCAC_R2.fq\n",
    "TCGACAAG_R1.fq\n",
    "TCGACAAG_R2.fq\n",
    "ATCATGCG_R1.fq                    \n",
    "ATCATGCG_R2.fq                    \n",
    "ATCGTGGT_R1.fq                     \n",
    "ATCGTGGT_R2.fq\n",
    "GATCTTGC_R1.fq\n",
    "GATCTTGC_R2.fq                                   \n",
    "CACTTCAC_R1.fq                     \n",
    "CACTTCAC_R2.fq                     \n",
    "CGATCGAT_R1.fq                      \n",
    "CGATCGAT_R2.fq \n",
    "TGTTCCGT_R1.fq         \n",
    "TGTTCCGT_R2.fq\n",
    "TCTTCGAC_R1.fq\n",
    "TCTTCGAC_R2.fq\n",
    "TCGGATTC_R1.fq\n",
    "TCGGATTC_R2.fq\n",
    "GTCCTAAG_R1.fq\n",
    "GTCCTAAG_R2.fq\n",
    "GTAGCGTA_R1.fq\n",
    "GTAGCGTA_R2.fq\n",
    "GCTACTCT_R1.fq \n",
    "GCTACTCT_R2.fq\n",
    "TCGAGAGT_R1.fq\n",
    "TCGAGAGT_R2.fq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test output \n",
    "head TACCGGAT_R1.fq\n",
    "12)Key in dictionary and matches, Good quality score, and match\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test output \n",
    "head TACCGGAT_R2.fq\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:ATCCGGTA\n",
    "12)Key in dictionary and matches, Good quality score, and match\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test output \n",
    "head AGGATAGC_R1.fq \n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:AGGATAGC\n",
    "9)Key in dictionary and matches \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:AGGATAGC\n",
    "10)Key in dictionary and matches\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test output \n",
    "head AGGATAGC_R2.fq\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:GCTATCCT\n",
    "9)Key in dictionary and matches \n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:GCTATCCT\n",
    "10)Key in dictionary and matches\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test output \n",
    "cat count_data\n",
    "# of Read Pairs with Low SQ R1\n",
    "3\n",
    "# of Read Pairs with Good SQ R1\n",
    "10\n",
    "# of Read Pairs with Ns\n",
    "2\n",
    "# of Read Pairs without Ns\n",
    "8\n",
    "# of Read Pairs Not in Dictionary\n",
    "4\n",
    "# of Read Pairs in Dictionary\n",
    "4\n",
    "# of Read Pairs that Don't Match\n",
    "1\n",
    "# of Read Pairs that Do Match\n",
    "3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "import argparse\n",
    "\n",
    "def get_arguments():\n",
    "    parser = argparse.ArgumentParser(description='k-mer size program')\n",
    "    parser.add_argument(\"-r1\", \"--read1\", help=\"input string\", required=True, type=str)\n",
    "    parser.add_argument(\"-r2\", \"--read2\", help=\"input string\", required=True, type=str)\n",
    "    parser.add_argument(\"-i1\", \"--index1\", help=\"input string\", required=True, type=str)\n",
    "    parser.add_argument(\"-i2\", \"--index2\", help=\"input string\", required=True, type=str)\n",
    "    parser.add_argument(\"-b\", \"--barcode_file\", help=\"input string\", required=True, type=str)\n",
    "    parser.add_argument(\"-c\", \"--cutoff\", help=\"input string\", required=True, type=str)\n",
    "    return parser.parse_args()\n",
    "    \n",
    "args = get_arguments()\n",
    "read1 = args.read1\n",
    "read2 = args.read2\n",
    "index1 = args.index1\n",
    "index2 = args.index2\n",
    "barcode_file = args.barcode_file\n",
    "cutoff = args.cutoff\n",
    "\n",
    "#actual code used\n",
    "\n",
    "#opening nano to make script\n",
    "#nano demultiplexing.py\n",
    "\n",
    "##########################################################################################\n",
    "#!/usr/bin/env python\n",
    "\n",
    "#Demultiplexing Python Assignment\n",
    "##################################################################################################################\n",
    "##Making Dictionary for Barcodes and output files\n",
    "\n",
    "#dictionary for R1 barcodes for making output file\n",
    "file_dict_r1={}\n",
    "#dictionary for R2 barcodes for making output files\n",
    "file_dict_r2={}\n",
    "#dictionary for checking if index matches\n",
    "barcode_dict = {}\n",
    "#Reading each line in the barcode file\n",
    "for line in open(barcode_file).readlines():\n",
    "    #striping the lines and spliting in the barcode file\n",
    "    sample, barcode = line.strip().split()\n",
    "    #setting the barcodes keys and the samples as values\n",
    "    barcode_dict[barcode] = sample\n",
    "    #creating R1 files that are named with all the barcodes in the dictionary\n",
    "    file_dict_r1[barcode]=open(barcode + \"_R1.fq\", \"w\")\n",
    "    #creating R2 files that are named with all the barcodes in the dictionary\n",
    "    file_dict_r2[barcode]=open(barcode + \"_R2.fq\", \"w\")\n",
    "\n",
    "\n",
    "##################################################################################################################\n",
    "##Creating Functions\n",
    "\n",
    "#making a function called convert_phred\n",
    "def convert_phred(letter):\n",
    "    #creating a variable that can be called later that converts the letter to a phred score\n",
    "    phred_score= ord(letter)-33\n",
    "    #returning the phred_score\n",
    "    return(phred_score)\n",
    "\n",
    "#making a function called reverse_complement\n",
    "def reverse_complement(bases):\n",
    "    #a dictionary that has the compliment of each base as values\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N':'N'}\n",
    "    #returning the compliment base and making the reverse.\n",
    "    return ''.join([complement[base] for base in bases[::-1]])\n",
    "\n",
    "##################################################################################################################\n",
    "#Creating Counters\n",
    "line_ctr=0\n",
    "\n",
    "\n",
    "\n",
    "ctr_low_qs=0\n",
    "ctr_good_qs=0\n",
    "\n",
    "ctr_yes_N=0\n",
    "ctr_No_N=0\n",
    "\n",
    "barcodes_notin_dict=0\n",
    "barcodes_in_dict=0\n",
    "\n",
    "\n",
    "reads_No_match=0\n",
    "reads_Do_match=0\n",
    "\n",
    "##################################################################################################################\n",
    "##Creating Variables for Files and Reading in Files\n",
    "#setting variables for each file\n",
    "\n",
    "#opening R1,R2,I1,I2 files and setting as variables\n",
    "#creating R1 and R2 undetermined files and setting as variables\n",
    "#creating a count data file and setting as variables\n",
    "with \\\n",
    "open(read1, \"r\") as r_1, open(read2, \"r\") as r_2, \\\n",
    "open(index1, \"r\") as i_1, open(index2, \"r\") as i_2, \\\n",
    "open('R1_undetermine.fq', 'w') as undetermined_r1, \\\n",
    "open('R2_undetermine.fq', 'w') as undetermined_r2, \\\n",
    "open('count_data', 'w') as count_data:\n",
    "\n",
    "##################################################################################################################\n",
    "##Creating Variables for Individual Lines\n",
    "    while r_1 and r_2 and i_1 and i_2:\n",
    "        #setting each line as a variable and stripping the new ling at the end\n",
    "        line1_i1 = i_1.readline().strip('\\n')\n",
    "        if not line1_i1:\n",
    "            break\n",
    "        line2_i1 = i_1.readline().strip('\\n')\n",
    "        line3_i1 = i_1.readline().strip('\\n')\n",
    "        line4_i1 = i_1.readline().strip('\\n')\n",
    "\n",
    "        #setting each line as a variable and stripping the new ling at the end\n",
    "        line1_i2 = i_2.readline().strip('\\n')\n",
    "        if not line1_i2:\n",
    "            break\n",
    "        line2_i2 = i_2.readline().strip('\\n')\n",
    "        line3_i2 = i_2.readline().strip('\\n')\n",
    "        line4_i2 = i_2.readline().strip('\\n')\n",
    "\n",
    "        #setting each line as a variable and stripping the new ling at the end\n",
    "        line1_r1 = r_1.readline().strip('\\n')\n",
    "        if not line1_r1:\n",
    "            break\n",
    "        line2_r1 = r_1.readline().strip('\\n')\n",
    "        line3_r1 = r_1.readline().strip('\\n')\n",
    "        line4_r1 = r_1.readline().strip('\\n')\n",
    "        #creating a variable that prints all 4 lines in fastq format with the index in the header\n",
    "        lines_r1 = line1_r1+\":\"+line2_i1+'\\n'+line2_r1+'\\n'+line3_r1+'\\n'+line4_r1+'\\n'\n",
    "        #print(lines_r1)\n",
    "\n",
    "        #setting each line as a variable and stripping the new ling at the end\n",
    "        line1_r2 = r_2.readline().strip('\\n')\n",
    "        if not line1_r2:\n",
    "            break\n",
    "        line2_r2 = r_2.readline().strip('\\n')\n",
    "        line3_r2 = r_2.readline().strip('\\n')\n",
    "        line4_r2 = r_2.readline().strip('\\n')\n",
    "        \n",
    "        rev_i2 = reverse_complement(line2_i2)\n",
    "        #creating a variable that prints all 4 lines in fastq format with the index in the header\n",
    "        lines_r2 = line1_r2+\":\"+rev_i2+'\\n'+line2_r2+'\\n'+line3_r2+'\\n'+line4_r2+'\\n'\n",
    "        #print(lines_r2)\n",
    "##################################################################################################################\n",
    "##Low Average Quality Score Removal\n",
    "#removing average read quality scores that are below a cutoff of 30. I choose 30 because all the quality\n",
    "##score were above 30 in the demultiplex index graphs, which any score above 30 is 99.9% accurate. In the\n",
    "##demultiplex index graphs, any cutoff above 30 will remove a good portion of good reads and we don't want\n",
    "##to do that. Low read quality scores are most likely due to Ns, which will be removed, so what will be left\n",
    "##is reads with average quality scores above 30 (99.9% accurate).\n",
    "\n",
    "\n",
    "        #creating a list to store characters in for index1\n",
    "        phred_list_i1=[]\n",
    "        #looking at each character in the quality score line for index1\n",
    "        for char in line4_i1:\n",
    "            #converting each character to a phred score\n",
    "            phred_score_i1=(convert_phred(char))\n",
    "            #print(phred_score_i1)#checking right characters are printed\n",
    "\n",
    "        #creating a list to store characters in for index2\n",
    "        phred_list_i2=[]\n",
    "        #looking at each character in the quality score line for index2\n",
    "        for char in line4_i2:\n",
    "            #converting each character to a phred score\n",
    "            phred_score_i2=(convert_phred(char))\n",
    "            #print(phred_score_i2)#checking right characters are printed\n",
    "\n",
    "            #appending characters to a list\n",
    "            phred_list_i1.append(phred_score_i1)\n",
    "            phred_list_i2.append(phred_score_i2)\n",
    "            #print(phred_list_i1)#making sure list look correctly\n",
    "            #print(phred_list_i2)#making sure list look correctly\n",
    "\n",
    "        #creating variables that has the sum of each line\n",
    "        phred_sum_i1 = sum(phred_list_i1)\n",
    "        phred_sum_i2 = sum(phred_list_i2)\n",
    "        #print(\"1\",phred_sum_i1)#checking to make sure sums are correct\n",
    "        #print(\"2\",phred_sum_i2)#checking to make sure sums are correct\n",
    "\n",
    "        #creating variables that are the average of each index line\n",
    "        phred_ave_i1 = (phred_sum_i1/len(line4_i1))\n",
    "        phred_ave_i2 = (phred_sum_i2/len(line4_i2))\n",
    "        #print(\"1\",phred_ave_i1)#checking to make sure average are correct\n",
    "        #print(\"2\",phred_ave_i2)#checking to make sure average are correct\n",
    "        #writing reads to undetermined file (R1 and R2) if they are below the set cutoff level.\n",
    "        if phred_ave_i1 < cutoff or phred_ave_i2 < cutoff:\n",
    "            #counting number of reads that are below the cutoff level\n",
    "            ctr_low_qs +=1\n",
    "            undetermined_r1.write(lines_r1)\n",
    "            undetermined_r2.write(lines_r2)\n",
    "        #if reads are above the cutoff level\n",
    "        else:\n",
    "            #counting the number of reads that are above the cutoff level\n",
    "            ctr_good_qs +=1\n",
    "#################################################################################################################\n",
    "##Removing Reads with Ns\n",
    "\n",
    "            #if index1 and index2 sequence lines have Ns. Reads are witten to undetermined files (R1 and R2)\n",
    "            if \"N\" in line2_i1 or \"N\" in line2_i2:\n",
    "                #counting number of reads that have Ns\n",
    "                ctr_yes_N +=1\n",
    "                undetermined_r1.write(lines_r1)\n",
    "                undetermined_r2.write(lines_r2)\n",
    "            else:\n",
    "#counting number of reads that don't have Ns\n",
    "                ctr_No_N +=1\n",
    "                #making the reverse compliment of index2 sequence\n",
    "                reverse_i2 = reverse_complement(line2_i2)\n",
    "                #print(\"original\", line2_i2)#viewing origional index2\n",
    "                #print(\"reverse\", reverse_i2)#making sure origional was reversed complimented\n",
    "                #finding indexes that are not in the dictionary and writing them to undertermined files (R1 and R2)\n",
    "                if line2_i1 not in barcode_dict.keys() or reverse_i2 not in barcode_dict.keys():\n",
    "                    #print(\"1\",line2_i1)#viewing index1\n",
    "                    #print(\"2\",reverse_i2)#making sure indexes are reversed complimented\n",
    "                    #counting the indexes that are not in the dictionary\n",
    "                    barcodes_notin_dict+=1\n",
    "                    undetermined_r1.write(lines_r1)\n",
    "                    undetermined_r2.write(lines_r2)\n",
    "\n",
    "                else:\n",
    "                    #finding the indexes that are in the dictionary\n",
    "                    if line2_i1 in barcode_dict.keys() and reverse_i2 in barcode_dict.keys():\n",
    "                        #counting the indexes that are in the dicitonary\n",
    "                        barcodes_in_dict += 1\n",
    "                        #print(\"R1\",line2_i1)#checking indexes that are found in the dictionary\n",
    "                        #print(\"R2\",reverse_i2)#checking indexes that are found in the dictionary\n",
    "\n",
    "                    #finind index reads that don't match and writing them to undertermined files (R1 and R2)\n",
    "                    if reverse_i2 != line2_i1:\n",
    "                        #counting index reads that don't match\n",
    "                        reads_No_match+=1\n",
    "                        #print(\"R1\",line2_i1)#checking indexes that don't match\n",
    "                        #print(\"R2\",reverse_i2)#checking indexes that don't match\n",
    "                        undetermined_r1.write(lines_r1)\n",
    "                        undetermined_r2.write(lines_r2)\n",
    "                        #finding indexes that match and if they do writing them to files that match their index sequence\n",
    "                    if reverse_i2==line2_i1:\n",
    "                        #counting reads that do match\n",
    "                        reads_Do_match+=1\n",
    "                        #print(\"R1\",line2_i1)#checking index reads that match\n",
    "                        #print(\"R2\",reverse_i2)#checking index reads that match\n",
    "                        #print(lines_r1)#checking associated reads that match\n",
    "                        #print(lines_r2)#checking associated reads that match\n",
    "                        file_dict_r1[line2_i1].write(lines_r1)\n",
    "                        file_dict_r2[reverse_i2].write(lines_r2)\n",
    "\n",
    "    #writing count data to a file called count_data\n",
    "    count_data.write(\"{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n\".format(\"# of Read Pairs with Low SQ R1\", ctr_low_qs, \\\n",
    "                                                               \"# of Read Pairs with Good SQ R1\", ctr_good_qs, \\\n",
    "                                                               \"# of Read Pairs with Ns\", ctr_yes_N, \\\n",
    "                                                               \"# of Read Pairs without Ns\", ctr_No_N, \\\n",
    "                                                               \"# of Read Pairs Not in Dictionary\", barcodes_notin_dict, \\\n",
    "                                                               \"# of Read Pairs in Dictionary\", barcodes_in_dict, \\\n",
    "                                                               \"# of Read Pairs that Don't Match\", reads_No_match, \\\n",
    "                                                               \"# of Read Pairs that Do Match\", reads_Do_match))\n",
    "\n",
    "#closing files that were created from the dictionary barcodes\n",
    "for barcode in file_dict_r1:\n",
    "    file_dict_r1[barcode].close()\n",
    "    file_dict_r2[barcode].close()\n",
    "\n",
    "###########################################################################################    \n",
    "#chaning file permissions to run the script    \n",
    "#chmod 755 demultiplexing.py\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#creating a batch script to run code on talapas\n",
    "nano demultiplexing.slurm\n",
    "\n",
    "###########################################################################################\n",
    "#!/usr/bin/env bash\n",
    "#SBATCH --partition=long    ### Partition\n",
    "#SBATCH --job-name=job  ### Job Name\n",
    "#SBATCH --output=job.out    ### File in which to store job output\n",
    "#SBATCH --time=2-00:00:00   ### Wall clock time limit in Days-HH:MM:SS\n",
    "#SBATCH --nodes=1           ### Number of nodes needed for the job\n",
    "#SBATCH --ntasks-per-node=20 ### Number of tasks to be launcged per Node\n",
    "\n",
    "cd /projects/bgmp/mdriskil/Bi624_Assignments/demultiplex\n",
    "\n",
    "./demultiplexing.py\n",
    "\n",
    "###########################################################################################\n",
    "#chaning file permission\n",
    "chmod 755 demultiplexing.slurm\n",
    "#submitting script to talapas\n",
    "sbatch demultiplexing.slurm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#actual output \n",
    "head R1_undetermine.fq \n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1:NCTTCGAC\n",
    "GNCTGGCATTCCCAGAGACATCAGTACCCAGTTGGTTCAGACAGTTCCTCTATTGGTTGACAAGGTCTTCATTTCTAGTGATATCAACACGGTGTCTACAA\n",
    "+\n",
    "A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F--JJJJJJF-A-F7JJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1286:1191 1:N:0:1:NACAGCGA\n",
    "CNACCTGTCCCCAGCTCACAGGACAGCACACCAAAGGCGGCAACCCACACCCAGTTTTACAGCCACACAGTGCCTTGTTTTACTTGAGGACCCCCCACTCC\n",
    "+\n",
    "A#AAFJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJAJJJJJJJJJJJJJJFJJJJJFFFFJJJJJJJJJJJJJJJJJJ77F\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 1:N:0:1:NTCCTAAG\n",
    "GNGGTCTTCTACCTTTCTCTTCTTTTTTGGAGGAGTAGAATGTTGAGAGTCAGCAGTAGCCTCATCATCACTAGATGGCATTTCTTCTGAGCAAAACAGGT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#actual output \n",
    "head R2_undetermine.fq \n",
    "@K00337:83:HJKJNBBXX:8:1101:1265:1191 4:N:0:1:TCTTCGAN\n",
    "NTTTTGATTTACCTTTCAGCCAATGAGAAGGCCGTTCATGCAGACTTTTTTAATGATTTTGAAGACCTTTTTGATGATGATGATGTCCAGTGAGGCCTCCC\n",
    "+\n",
    "#AAFAFJJ-----F---7-<FA-F<AFFA-JJJ77<FJFJFJJJJJJJJJJAFJFFAJJJJJJJJFJF7-AFFJJ7F7JFJJFJ7FFF--A<A7<-A-7--\n",
    "@K00337:83:HJKJNBBXX:8:1101:1286:1191 4:N:0:1:AACAGCGN\n",
    "NTGTGTAGACAAAAGTTTTCATGAGTCTGTAAGCTGTCTATTGTCTCCTGAAAAGAAACCAGAAGTTTTCCCCTAAATGTGTTTAGAATGCTTATTCTAAT\n",
    "+\n",
    "#A-AFFJJFJJJJJJJJJJJJJJJJ<JAJFJJJJF<JFJJJAJJJJJJJJJJJJJJJJJJJFJJJAJJFJJJFJJJF<JJA-JJJ-<AFAF--FF<JAFJF\n",
    "@K00337:83:HJKJNBBXX:8:1101:1347:1191 4:N:0:1:GTCCTAAN\n",
    "NAAATGCCATCTAGTGATGATGAGGCTACTGCTGACTCTCAACATTCTACTCCTCCAAAAAAGAAGAGAAAGATTCCAACCCCCAGAACCGATGACCGGCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#actual output \n",
    "head AACAGCGA_R1.fq\n",
    "@K00337:83:HJKJNBBXX:8:1101:1479:1701 1:N:0:1:AACAGCGA\n",
    "CCTTTGCTTAATCTGCTCTTGTCTTTCAGCACTAAGCTGTTCTTTTTCTTCTTTCATGGCTGAAATTTTTGTTTTCAGTTCTCTAACCTGGCGCTCAATAT\n",
    "+\n",
    "AAAAAJJFJJJJJJJJJJJJFJJJJFJJJJJJFJJJJJFFJJJJJJJJJJJJJJFJJJJJJJAFJJJJJJJJJJJJJFF<JJJJJJJJJJJFJJJJJFJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1966:1701 1:N:0:1:AACAGCGA\n",
    "ATGTTCAGTTGTCTCTGGACTTACAAGTTCCTCAGTACACTGCTCCAGAGCCTCCCTCTCTTCATCCATGTCCTGACTGGTCTCCTTGCTGGCTTCTGAGC\n",
    "+\n",
    "AAFAFJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJFJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:4076:1701 1:N:0:1:AACAGCGA\n",
    "TGCCCCCTTCCTGGGCATCCGTTCGGGTTTTCTAACCGTTGTCTTGTTCTGGGGTGAGCGGGGAGCGCATGCAGAGATCCCAAGGCCTATGTCTGGAGGGT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#actual output \n",
    "head AACAGCGA_R2.fq\n",
    "@K00337:83:HJKJNBBXX:8:1101:1479:1701 4:N:0:1:AACAGCGA\n",
    "CAACGAGACGCGCGCTAAGCTCGACGAGCTTTCTGCTAAGCGAGAAACGAGTGGAGAGAAATCCAGACAACTAAGGGATGCCCAGCAGGATGCAAGAGACA\n",
    "+\n",
    "A<AFFJJJJJJJJJJJJJJJAJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJ7JJJJJJJJJJJJJJJJJJFJJJJJJJJJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:1966:1701 4:N:0:1:AACAGCGA\n",
    "CTTGTGAGAACTGAACCACGAGAGCAGCTCCTTGATCAGCTGCAAAAGAAACAACCACCAATGATGCTAAACAGCTCAGAAGCCAGCAAGGAGACCAGTCA\n",
    "+\n",
    "AAFFFJJJAJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJFJJ\n",
    "@K00337:83:HJKJNBBXX:8:1101:4076:1701 4:N:0:1:AACAGCGA\n",
    "CCGAGCGGCAGAGAGTGGGGTAGGCAGGAGGCTCCCTGTAAACATTTGGACTTGGGTCAGGGCTGGTGTTGGACAAAGCCAGGGGTCTAGGCTGGAGGAGT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#number of reads in original files\n",
    "1294_S1_L008_R1_001.fastq | wc -l\n",
    "1452986940/4=363246735\n",
    "cat 1294_S1_L008_R2_001.fastq | wc -l\n",
    "1452986940/4=363246735\n",
    "cat 1294_S1_L008_R3_001.fastq | wc -l\n",
    "1452986940/4=363246735\n",
    "cat 1294_S1_L008_R4_001.fastq | wc -l\n",
    "1452986940/4=363246735\n",
    "\n",
    "#number of reads in undetermined files\n",
    "cat R1_undetermine.fq | wc -l\n",
    "239463044/4=59865761 \n",
    "cat R2_undetermine.fq | wc -l\n",
    "239463044/4=59865761\n",
    "59865761/363246735=0.1648074 proportion\n",
    "\n",
    "#number of reads in 48 fastq files\n",
    "cat AACAGCGA_R1.fq | wc -l\n",
    "32356964/4=8089241\n",
    " cat AACAGCGA_R2.fq | wc -l\n",
    "32356964/4=8089241\n",
    "8089241/363246735=0.02226927 proportion\n",
    "\n",
    "\n",
    "cat ACGATCAG_R1.fq | wc -l\n",
    "30192604/4=7548151\n",
    "cat ACGATCAG_R2.fq | wc -l\n",
    "30192604/4=7548151\n",
    "7548151/363246735=0.02077968 proportion\n",
    "\n",
    "cat AGAGTCCA_R1.fq | wc -l\n",
    "42039364/4=10509841\n",
    "cat AGAGTCCA_R2.fq | wc -l\n",
    "42039364/4=10509841\n",
    "10509841/363246735=0.02893306 proportion\n",
    "\n",
    "cat AGGATAGC_R1.fq | wc -l\n",
    "32629256/4=8157314\n",
    "cat AGGATAGC_R2.fq | wc -l\n",
    "32629256/4=8157314\n",
    "8157314/363246735=0.02245668 proportion\n",
    "\n",
    "cat ATCATGCG_R1.fq | wc -l\n",
    "37543456/4=9385864\n",
    "cat ATCATGCG_R2.fq | wc -l\n",
    "37543456/4=9385864\n",
    "9385864/363246735=0.02583881 proportion\n",
    "\n",
    "cat ATCGTGGT_R1.fq | wc -l\n",
    "24685116/4=6171279\n",
    "cat ATCGTGGT_R2.fq | wc -l\n",
    "24685116/4=6171279\n",
    "6171279/363246735=0.01698922 proportion\n",
    "\n",
    "cat CACTTCAC_R1.fq | wc -l\n",
    "15547792/4=3886948\n",
    "cat CACTTCAC_R2.fq | wc -l\n",
    "15547792/4=3886948\n",
    "3886948/363246735=0.01070057 proportion\n",
    "\n",
    "cat CGATCGAT_R1.fq | wc -l\n",
    "20812540/4=5203135\n",
    "cat CGATCGAT_R2.fq | wc -l\n",
    "20812540/4=5203135\n",
    "5203135/363246735=0.01432397 proportion\n",
    "\n",
    "cat CGGTAATC_R1.fq | wc -l\n",
    "18208596/4=4552149\n",
    "cat CGGTAATC_R2.fq | wc -l\n",
    "18208596/4=4552149\n",
    "4552149/363246735=0.01253184 proportion\n",
    "\n",
    "cat CTAGCTCA_R1.fq | wc -l\n",
    "65049000/4=16262250\n",
    "cat CTAGCTCA_R2.fq | wc -l\n",
    "650490004=16262250\n",
    "16262250/363246735=0.04476916 proportion\n",
    "\n",
    "cat CTCTGGAT_R1.fq | wc -l\n",
    "127253488/4=31813372\n",
    "cat CTCTGGAT_R2.fq | wc -l\n",
    "127253488/4=31813372\n",
    "31813372/363246735=0.08758061 proportion\n",
    "\n",
    "cat GATCAAGG_R1.fq | wc -l\n",
    "24661800/4=6165450\n",
    "cat GATCAAGG_R2.fq | wc -l\n",
    "24661800/4=6165450\n",
    "6165450/363246735=0.01697317 proportion\n",
    "\n",
    "cat GATCTTGC_R1.fq | wc -l\n",
    "13813748/4=3453437\n",
    "cat GATCTTGC_R2.fq | wc -l\n",
    "13813748/4=3453437\n",
    "3453437/363246735=0.009507138 proportion\n",
    "\n",
    "GCTACTCT_R1.fq | wc -l\n",
    "26461564/4=6615391\n",
    "cat GCTACTCT_R2.fq | wc -l\n",
    "26461564/4=6615391\n",
    "6615391/363246735=0.01821184 proportion\n",
    "\n",
    "cat GTAGCGTA_R1.fq | wc -l\n",
    "29790372/4=7447593\n",
    "cat GTAGCGTA_R2.fq | wc -l\n",
    "29790372/4=7447593\n",
    "7447593/363246735=0.02050285 proportion\n",
    "\n",
    "cat GTCCTAAG_R1.fq | wc -l\n",
    "32994664/4=8248666\n",
    "cat GTCCTAAG_R2.fq | wc -l\n",
    "32994664/4=8248666\n",
    "8248666/363246735=0.02270816 proportion\n",
    "\n",
    "cat TACCGGAT_R1.fq | wc -l\n",
    "272873952/4=68218488\n",
    "cat TACCGGAT_R2.fq | wc -l\n",
    "272873952/4=68218488\n",
    "68218488/363246735=0.1878021 proportion\n",
    "\n",
    "cat TAGCCATG_R1.fq | wc -l\n",
    "39729840/4=9932460\n",
    "cat TAGCCATG_R2.fq | wc -l\n",
    "39729840/4=9932460\n",
    "9932460/363246735=0.02734356 proportion\n",
    "\n",
    "cat TATGGCAC_R1.fq | wc -l\n",
    "41122396/4=10280599\n",
    "cat TATGGCAC_R2.fq | wc -l\n",
    "41122396/4=10280599\n",
    "10280599/363246735=0.02830197 proportion\n",
    "\n",
    "cat TCGACAAG_R1.fq | wc -l\n",
    "14295448/4=3573862\n",
    "cat TCGACAAG_R2.fq | wc -l\n",
    "14295448/4=3573862\n",
    "3573862/363246735=0.009838662 proportion\n",
    "\n",
    "cat TCGAGAGT_R1.fq | wc -l\n",
    "38668032/4=9667008\n",
    "cat TCGAGAGT_R2.fq | wc -l\n",
    "38668032/4=9667008\n",
    "9667008/363246735=0.02661279 proportion\n",
    "\n",
    "cat TCGGATTC_R1.fq | wc -l\n",
    "17165336/4=4291334\n",
    "cat TCGGATTC_R2.fq | wc -l\n",
    "17165336/4=4291334\n",
    "4291334/363246735=0.01181383 proportion\n",
    "\n",
    "cat TCTTCGAC_R1.fq | wc -l\n",
    "157519852/4=39379963\n",
    "cat TCTTCGAC_R2.fq | wc -l\n",
    "157519852/4=39379963\n",
    "39379963/363246735=0.1084111 proportion\n",
    "\n",
    "cat TGTTCCGT_R1.fq | wc -l\n",
    "58108716/4=14527179\n",
    "cat TGTTCCGT_R2.fq | wc -l\n",
    "58108716/4=14527179\n",
    "14527179/363246735=0.03999259 proportion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat count_data \n",
    "# of Read Pairs with Low SQ R1\n",
    "46324005\n",
    "46324005/363246735=0.1275277 proportion\n",
    "\n",
    "# of Read Pairs with Good SQ R1\n",
    "316922730\n",
    "316922730/363246735=0.8724723 proportion\n",
    "\n",
    "# of Read Pairs with Ns\n",
    "3139834\n",
    "3139834/363246735=0.008643805 proportion\n",
    "\n",
    "# of Read Pairs without Ns\n",
    "313782896\n",
    "313782896/363246735=0.8638285 proportion\n",
    "\n",
    "# of Read Pairs Not in Dictionary\n",
    "9866439\n",
    "9866439/363246735=0.02716181 proportion\n",
    "\n",
    "# of Read Pairs in Dictionary\n",
    "303916457\n",
    "303916457/363246735=0.8366667 proportion\n",
    "\n",
    "# of Read Pairs that Don't Match-index hopping\n",
    "535483\n",
    "535483/363246735=0.001474158 proportion\n",
    "\n",
    "# of Read Pairs that Do Match\n",
    "303380974\n",
    "303380974/363246735=0.8351926 proportion\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
